kasa/discover.py

```diff
@@ -37,86 +108,76 @@ class Discover:
-        :param timeout: How long to wait for responses, defaults to 3
-        :param port: port to send broadcast messages, defaults to 9999.
-        :rtype: dict
-        :return: Array of json objects {"ip", "port", "sys_info"}
+        :param on_discovered:
+        :param timeout: How long to wait for responses, defaults to 5
+        :param discovery_packets: Number of discovery packets are broadcasted.
+        :param return_raw: True to return JSON objects instead of Devices.
+        :return:
         """
-        if protocol is None:
-            protocol = TPLinkSmartHomeProtocol()
+        loop = asyncio.get_event_loop()
+        transport, protocol = await loop.create_datagram_endpoint(
+            lambda: _DiscoverProtocol(
+                target=target,
+                on_discovered=on_discovered,
+                timeout=timeout,
+                discovery_packets=discovery_packets,
+            ),
+            local_addr=("0.0.0.0", 0),
+        )
+        protocol = cast(_DiscoverProtocol, protocol)
 
-        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
-        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        sock.settimeout(timeout)
-
-        req = json.dumps(Discover.DISCOVERY_QUERY)
-        _LOGGER.debug("Sending discovery to %s:%s", target, port)
+        try:
+            _LOGGER.debug("Waiting %s seconds for responses...", timeout)
+            await asyncio.sleep(5)
+        finally:
+            transport.close()
 
-        encrypted_req = protocol.encrypt(req)
-        for i in range(discovery_packets):
-            sock.sendto(encrypted_req[4:], (target, port))
+        _LOGGER.debug("Discovered %s devices", len(protocol.discovered_devices))
 
-        devices = {}
-        _LOGGER.debug("Waiting %s seconds for responses...", timeout)
+        if return_raw:
+            return protocol.discovered_devices_raw
 
-        try:
-            while True:
-                data, addr = sock.recvfrom(4096)
-                ip, port = addr
-                info = json.loads(protocol.decrypt(data))
-                device_class = Discover._get_device_class(info)
-                if return_raw:
-                    devices[ip] = info
-                elif device_class is not None:
-                    devices[ip] = device_class(ip)
-        except socket.timeout:
-            _LOGGER.debug("Got socket timeout, which is okay.")
-        except Exception as ex:
-            _LOGGER.error("Got exception %s", ex, exc_info=True)
-        _LOGGER.debug("Found %s devices: %s", len(devices), devices)
-        return devices
+        return protocol.discovered_devices
 
     @staticmethod
-    async def discover_single(
-        host: str, protocol: TPLinkSmartHomeProtocol = None
-    ) -> Optional[SmartDevice]:
+    async def discover_single(host: str) -> Optional[SmartDevice]:
```

Comment:
I think this shouldn't be `Optional`. It returns a `SmartDevice` or it raises IIUC.

Reply:
Indeed! I converted to use exceptions instead of returning `None`, as I think that's nicer API-wise for the users; either the exception bubbles up or gets caught, and there's no need for weird `if x is not None` structures. I'll update it.